// @ampt/sdk/dist/chunk-UQBSQAXE.js
import { createRequire as topLevelCreateRequire } from 'module'
const require = topLevelCreateRequire(import.meta.url || __dirname)
import {
  __commonJS,
  __toESM
} from "./chunk-TTZTDAPF.js";

// ../../node_modules/ramda/src/internal/_isPlaceholder.js
var require_isPlaceholder = __commonJS({
  "../../node_modules/ramda/src/internal/_isPlaceholder.js"(exports, module) {
    function _isPlaceholder(a) {
      return a != null && typeof a === "object" && a["@@functional/placeholder"] === true;
    }
    module.exports = _isPlaceholder;
  }
});

// ../../node_modules/ramda/src/internal/_curry1.js
var require_curry1 = __commonJS({
  "../../node_modules/ramda/src/internal/_curry1.js"(exports, module) {
    var _isPlaceholder = require_isPlaceholder();
    function _curry1(fn) {
      return function f1(a) {
        if (arguments.length === 0 || _isPlaceholder(a)) {
          return f1;
        } else {
          return fn.apply(this, arguments);
        }
      };
    }
    module.exports = _curry1;
  }
});

// ../../node_modules/ramda/src/internal/_curry2.js
var require_curry2 = __commonJS({
  "../../node_modules/ramda/src/internal/_curry2.js"(exports, module) {
    var _curry1 = require_curry1();
    var _isPlaceholder = require_isPlaceholder();
    function _curry2(fn) {
      return function f2(a, b) {
        switch (arguments.length) {
          case 0:
            return f2;
          case 1:
            return _isPlaceholder(a) ? f2 : _curry1(function(_b) {
              return fn(a, _b);
            });
          default:
            return _isPlaceholder(a) && _isPlaceholder(b) ? f2 : _isPlaceholder(a) ? _curry1(function(_a) {
              return fn(_a, b);
            }) : _isPlaceholder(b) ? _curry1(function(_b) {
              return fn(a, _b);
            }) : fn(a, b);
        }
      };
    }
    module.exports = _curry2;
  }
});

// ../../node_modules/ramda/src/omit.js
var require_omit = __commonJS({
  "../../node_modules/ramda/src/omit.js"(exports, module) {
    var _curry2 = require_curry2();
    var omit2 = /* @__PURE__ */ _curry2(function omit3(names, obj) {
      var result = {};
      var index = {};
      var idx = 0;
      var len = names.length;
      while (idx < len) {
        index[names[idx]] = 1;
        idx += 1;
      }
      for (var prop in obj) {
        if (!index.hasOwnProperty(prop)) {
          result[prop] = obj[prop];
        }
      }
      return result;
    });
    module.exports = omit2;
  }
});

// ../../lib/ampt-utilities/src/params.ts
var import_omit = __toESM(require_omit(), 1);
var disallowedKeys = ["AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY", "AWS_SESSION_TOKEN"];
function parseParams() {
  try {
    return JSON.parse(process.env.PARAMS || "{}");
  } catch (err) {
    return {};
  }
}
function filterDisallowedExportParams(params2) {
  return (0, import_omit.default)(disallowedKeys, params2);
}

// params/index.ts
var existingEnv = {
  ORG_NAME: process.env.AMPT_ORG_NAME,
  SERVICE_NAME: process.env.AMPT_SERVICE_NAME,
  APP_NAME: process.env.AMPT_SERVICE_NAME,
  INSTANCE_NAME: process.env.AMPT_INSTANCE_NAME,
  ENVIRONMENT_NAME: process.env.AMPT_INSTANCE_NAME,
  ENV_NAME: process.env.AMPT_INSTANCE_NAME,
  INSTANCE_TYPE: process.env.AMPT_INSTANCE_TYPE,
  ENVIRONMENT_TYPE: process.env.AMPT_INSTANCE_TYPE,
  ENV_TYPE: process.env.AMPT_INSTANCE_TYPE,
  STAGE_NAME: process.env.AMPT_STAGE,
  AMPT_URL: process.env.AMPT_URL,
  AMPT_REGION: process.env.AWS_REGION
};
var Params = class {
  params;
  constructor() {
    this.params = {
      ...parseParams(),
      ...existingEnv
    };
  }
  export(keys = []) {
    const allowedParams = filterDisallowedExportParams(this.params);
    if (!keys.length) {
      process.env = {
        ...allowedParams,
        ...process.env
      };
    } else {
      keys.forEach((key) => {
        if (!process.env[key] && allowedParams[key]) {
          process.env[key] = allowedParams[key];
        }
      });
    }
  }
  list() {
    return this.params;
  }
  handler(key) {
    if (key) {
      return this.params[key] || void 0;
    }
    return {
      export: this.export.bind(this),
      list: this.list.bind(this)
    };
  }
};
var _params = new Params();
var params = _params.handler.bind(_params);

export {
  Params,
  params
};
